### PR TITLE
Improve Textfield component

### LINKED ISSUE #188: Improve Textfield component
The ```Textfield``` component is missing a lot of features. For reference, [this is the GitHub documentation](https://github.com/material-components/material-components-web/tree/master/packages/mdc-textfield).

Missing features include:
- [ ] ```mdc-textfield--disabled``` class on disabled fields
- [ ] Dense
- [ ] Helper text (that can be persistent or not)
- [ ] The label is only floating when a value is set on the field, but it should float on focus

### PR DESCRIPTION
**Pull request by filoozom**
Maybe shouldn't have created an issue (fixes #188), but didn't know I'd have time to work on this straight away. This is a work in progress to add all the missing functionality to ```Textfield```.

- [x] Add ```mdc-textfield--disabled``` class on disabled fields
- [x] Add dense
- [x] The label is only floating when a value is set on the field, but it should float on focus
- [x] Add helper text (that can be persistent or not)
- [x] Add text field boxes
- [x] Avoid Flash Of Un-styled Content (FOUC)
- [x] CSS only textfield

I think we should also implement some way to recursively remove all ```_mdcProps``` from child elements. For example, in ```Textfield``` the ```<input>``` or ```<textarea>``` get all ```{...props}```, and the ```_mdcProps``` aren't removed from them because they aren't the parent element, if I understand it correctly.

### CODE REVIEW

**Comment by filoozom** [comment]
So, I implemented a basic helper text, but there are two problems:

1. We need to set an ID to the component: ```<Textfield helptext="Text" id="id" />``` because we need it for the ```aria-controls``` on the ```input``` / ```textarea``` and on the ```<Helptext>```.
2. We have an useless outer ```div``` when ```helptext``` is not used.
3. The outer ```div``` has an ```mdc-``` class because of ```MaterialComponent```

Potential fixes:
1. I thaught about generating random IDs, but is that really a good idea?
2. I would have like for the ```Helptext``` to be an exported component that one could put independently from a ```Textfield``` because I feel like it would be a lot cleaner, both for props and for ```Textfield```'s ```materialDom``` function, but I don't know if it would work because I think that ```componentDidMount``` from the ```Textfield``` might fire before the ```Helptext``` component renders, which means that the ```Helptext``` would never be linked to the input as per https://github.com/material-components/material-components-web/blob/master/packages/mdc-textfield/index.js#L36.
3. Maybe we should not add ```mdc-``` if ```componentName``` is empty.

Any ideas @prateekbh?

**Comment by filoozom** [comment]
Ended up implementing something similar to https://github.com/developit/preact-mdl/blob/master/src/index.js#L15-L19 / https://github.com/developit/preact-mdl/blob/master/src/index.js#L854 for the automatic IDs.

**Comment by prateekbh** [comment]
All of your recent work has been amazing!!! 💯 💯 💯 
I will merge this and publish a new version soon. Will you be awesome/kind enough to update `gh-pages` as well?

**Comment by filoozom** [comment]
I will for sure 😃

File: MaterialComponent.js [hunk 1]
```diff
@@ -20,6 +20,12 @@ export default class MaterialComponent extends Component {
     // The final class name given to the dom
     this.classText = "";
   }
+  static uid() {
```
Comment (prateekbh) [inline MaterialComponent.js]: can we move this to textfield instead, cuz no other component use this

Comment (filoozom) [inline MaterialComponent.js]: Sure. Thought I'd put it there for future-proofing because other elements might need it. But for now it's probably better to keep it in ```Textfield```.

File: Textfield/Textfield.jsx [hunk 2]
```diff
@@ -3,49 +3,120 @@ import MaterialComponent from "../MaterialComponent";
...
     super();
     this.componentName = "textfield";
-    this.state = {
-      showFloatingLabel: false
-    };
-    this._mdcProps = ["fullwidth", "multiline"];
+    this._mdcProps = ["fullwidth", "multiline", "dense", "disabled"];
   }
   componentDidMount() {
-    this.setState({
```
Comment (prateekbh) [inline Textfield/Textfield.jsx]: this check was in place because floating labels(the way we are providing them) are only possible for JS enabled clients.

for a non jS client they should be outside the container

https://github.com/material-components/material-components-web/tree/master/packages/mdc-textfield#multi-line---gracefully-degraded

File: Textfield/Textfield.jsx [hunk 3]
```diff
@@ -3,49 +3,120 @@ import MaterialComponent from "../MaterialComponent";
...
+    }
+    return ++this.uidCounter;
+  }
+
+  render(allprops) {
+    const {
+      id,
+      className,
+      helptextPersistent,
+      helptextValidationMsg,
```
Comment (prateekbh) [inline Textfield/Textfield.jsx]: would it be better to have full `helptextValidationMessage`?

Comment (filoozom) [inline Textfield/Textfield.jsx]: As you prefer, I chose ```helptextValidationMsg``` because of the name of the class, which is ```mdc-textfield-helptext-validation-msg```, so it was just to stay as close to original as possible, but I can change it if you want.

Comment (prateekbh) [inline Textfield/Textfield.jsx]: oh no, it makes sense now. lets keep it this way

File: Textfield/Textfield.jsx (hunk 2 above)
Comment (prateekbh) [inline Textfield/Textfield.jsx]: please revert this change

Comment (filoozom) [inline Textfield/Textfield.jsx]: I'm not at home right now, but I'm pretty sure that it didn't work before either. The documentation is broken right now for example, isn't it? I'll have a look at it in a few hours.

Comment (prateekbh) [inline Textfield/Textfield.jsx]: sure no probs, so the idea is: `componentDidMount` executes only on client, hence when jS is present hence the check

Comment (filoozom) [inline Textfield/Textfield.jsx]: Right, I got that, but the issue is that the state in ```componentDidMount``` is changed after ```this.MDComponent``` is instanciated, which means that the label only appears after ```MDCTextfield``` has set the label to ```null``` because it didn't exist, which means that ```material-components-web``` doesn't update the label and add the floating class on focus.
I'm on the phone, not quite sure if what I'm writing makes sense, but I'm pretty sure that it didn't work before. Just check on the documentation, you'll see that the label never floats up.

**Comment by filoozom** [comment]
I think this should work. Now we just call ```this.MDComponent = new MDCTextfield(this.control);``` once the rerender with the label is done, which is the important part!

**Comment by filoozom** [comment]
I now have somewhat of a working example that works for both CSS-only and JS. Also added ```cssLabel``` to set a different label when it's CSS-only. I think this might be useful for other languages for example. I now wonder if I shouldn't just add a ```Textfield.CSSLabel``` instead though. Maybe even a ```no-js``` class or something like that.
Just need to tidy it up somehow for the ```showFloatingLabel``` state. Do you have any ideas?

File: Textfield/Textfield.jsx [hunk 4]
```diff
@@ -1,50 +1,156 @@
...
+    const tfId = id || "tf-" + this.id;
+
+    // Helper text
+    const helptextProps = {
+      persistent: helptextPersistent,
+      "validation-msg": helptextValidationMsg
+    };
+
+    return (
+      <div className={className}>
```
Comment (prateekbh) [inline Textfield/Textfield.jsx]: this will result into div->div->input which breaks us away from mdc-web
put the `{showFloatingLabel &&` check outside and put the extra div only incase of no js

File: Textfield/Textfield.jsx [hunk 5]
```diff
@@ -1,50 +1,156 @@
...
+        {props.label &&
+          !showFloatingLabel &&
+          <label for={tfId}>
+            {props.cssLabel || `${props.label}: `}
           </label>}
+        <TextfieldInput {...props} id={tfId} />
```
Comment (prateekbh) [inline Textfield/Textfield.jsx]: incase I dont want helper text or anything else, `className` should reach to the input tag

File: Textfield/Textfield.jsx (hunk 4 above)
Comment (filoozom) [inline Textfield/Textfield.jsx]: But also when there is a help text then I guess.

**Comment by prateekbh** [comment]
>One other somewhat "breaking" change is that Textfield is now always inside a <div> that gets the className instead of the TextfieldInput that did before. But this actually makes it more consistent with the labels and text helpers.

we should make the div has something like `className-container`.

lets make sure that end result markup should be same as mdc-web

**Comment by filoozom** [comment]
Ok, so now I added ```Textfield.Helptext```, so that we can do something like:
```html
<Textfield label="Username" id="textfield" />
<Textfield.Helptext id="textfield-helptext">
	Test
</Textfield.Helptext>
```
but the ```id``` would have to be set accordingly. At least like that we don't have the outer ```<div>```. What do you prefer?

File: Textfield/Textfield.jsx [hunk 6]
```diff
@@ -1,51 +1,176 @@
...
+            <label for={tfId}>
+              {props.cssLabel || `${props.label}: `}
+            </label>}
+          <TextfieldInput {...props} id={tfId} />
+          {props.helptext &&
+            <Helptext id={tfId + "-helptext"} {...helptextProps}>
+              {props.helptext}
+            </Helptext>}
+        </div>
+      : <TextfieldInput {...props} id={tfId} className={className} />;
```
Comment (prateekbh) [inline Textfield/Textfield.jsx]: lets not put auto generated Id for this case

Comment (filoozom) [inline Textfield/Textfield.jsx]: Right, that one didn't quite make sense, my bad.

**Comment by filoozom** [comment]
Do you have a more elegant solution to replace the double
```
this.setState({
      showFloatingLabel: true
    });

File: Textfield/Textfield.jsx [hunk 7]
```diff
@@ -167,7 +167,7 @@ class Textfield extends Component {
               {props.helptext}
             </Helptext>}
         </div>
-      : <TextfieldInput {...props} id={tfId} className={className} />;
```
Comment (prateekbh) [inline Textfield/Textfield.jsx]: since you deconstructed `id` props wont have it. and now even if I give an id it wont be applied.

better would be to put it back in props I guess

Comment (filoozom) [inline Textfield/Textfield.jsx]: My bad, it's early. Will fix this as soon as I get my hands on a computer.

**Comment by filoozom** [comment]
Well, I guess I should've put ```<TextfieldInput {...props} className={className} id={id} />``` for consistency. Or do you prefer something like:
```js
if (!showDiv) {
  props.id = id;
}
```
?

**Comment by prateekbh** [comment]
>```
if (!showDiv) {
	props.id = id;
}
```

☝️ is preferred.

yeah, `var canUseDOM = !!(` can actually be better

**Comment by filoozom** [comment]
Should we use it then? Or should we keep the state for now and maybe move to that one later, maybe after testing is done to ensure that everything works well?

**Comment by prateekbh** [comment]
we should keep things as is now and come back later

**Comment by filoozom** [comment]
Changed a few things again, hope I didn't messed up and that it'll do it 😛 Found a small bug where it added ```aria-controls``` without an helper text, so removed that.
I feel like ```const showDiv = props.helptext || (props.label && !showFloatingLabel);``` was changed to ```const showDiv = props.helptext || props.label & !showFloatingLabel;``` automatically which didn't work, or maybe I need a lot more sleep than I thought I did...

File: Textfield/Textfield.jsx [hunk 8]
```diff
@@ -1,51 +1,175 @@
...
+    this.MDComponent && this.MDComponent.destroy && this.MDComponent.destroy();
   }
   materialDom(allprops) {
-    const { className, ...props } = allprops;
-    let labelClass = ["mdc-textfield__label"];
-    if (props.value) {
-      labelClass.push("mdc-textfield__label--float-above");
+    let { className, ...props } = allprops;
+
+    if (props.value && this.state.showFloatingLabel) {
```
Comment (prateekbh) [inline Textfield/Textfield.jsx]: change this to 'value' in props. that will fix the null state problem
